### PR TITLE
Id 332 user service spec

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
Ticket: <Link to Jira ticket>

  \<Don't forget to include the ticket number in the PR title!\>

What:

  \<For your reviewers' sake, please describe in a sentence or two what this PR is accomplishing (usually from the users' perspective, but not necessarily).\>

Why:

  \<For your reviewers' sake, please describe in ~1 paragraph what the value of this PR is to our users or to ourselves.\>

How:

  \<For your reviewers' sake, please describe in ~1 paragraph how this PR accomplishes its goal.\>

  \<If the PR is big, please indicate where a reviewer should start reading it (i.e. which file or function).\>

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpec.scala [hunk 1]
```diff
@@ -36,167 +48,251 @@ class UserServiceSpec
...
+    mockTosService = mock[TosService](RETURNS_SMART_NULLS)
+    when(mockTosService.getTosStatus(any[WorkbenchUserId], any[SamRequestContext])).thenReturn(IO(Option(true)))
+
+    service = Mockito.spy(new UserService(dirDAO, googleExtensions, Seq(blockedDomain), mockTosService))
   }
 
-  protected def clearDatabase(): Unit =
-    TestSupport.truncateAll
+  // TODO: Add nested suites for each of the component test suites:
+  // https://www.scalatest.org/scaladoc/3.2.3/org/scalatest/Suite.html
```
Comment (Ghost-in-a-Jar) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpec.scala]: looks like maybe you did this already?

Comment (gpolumbo-broad) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpec.scala]: My intention here is to leave this comment as a call to action to keep extracting tests into their own "more focused" suites and then call the nested suites from the new `UserServiceSpec`

File: src/test/resources/logback-test.xml [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
     <if condition='!isDefined("SAM_LOG_APPENDER")'>
         <then>
-            <variable name="SAM_LOG_APPENDER" value="Console-Stackdriver" />
+            <variable name="SAM_LOG_APPENDER" value="Console-Standard" />
```
Comment (gpolumbo-broad) [inline src/test/resources/logback-test.xml:6]: Changes in `logback-test.xml` keep the log appender from spewing out lots of stackdriver trace information during tests

**Comment by gpolumbo-broad** [comment]
jenkins retest this when you have a second bud

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/MockDirectoryDaoBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      .loadGroup(ArgumentMatchers.eq(WorkbenchGroupName(allUsersGroup.id.toString)), any[SamRequestContext])
+
+    doReturn(IO(true))
+      .when(mockedDirectoryDAO)
+      .addGroupMember(ArgumentMatchers.eq(allUsersGroup.id), any[WorkbenchSubject], any[SamRequestContext])
+
+    this
+  }
+
+  // Bare minimum for a user to exist:
```
Comment (Shakespeared) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/MockDirectoryDaoBuilder.scala:120]: these comments are super helpful

**Review by Shakespeared — APPROVED** [review]
I didn't do a super close walkthrough of every test, but LGTM overall

**Review by Ghost-in-a-Jar — APPROVED** [review]
this looks great, only detail that maybe could be even clearer is to separate the mock setup in the test suites into functions -- might make it easier to understand what they are doing at a glance. I think that could be in a followup pr though.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpec.scala [hunk 4]
```diff
@@ -37,167 +53,318 @@ class UserServiceSpec
...
+    when(dirDAO.deleteUser(defaultUser.id, samRequestContext)).thenReturn(IO(()))
+    when(dirDAO.enableIdentity(defaultUser.id, samRequestContext)).thenReturn(IO(()))
+    when(dirDAO.disableIdentity(defaultUser.id, samRequestContext)).thenReturn(IO(()))
+    when(dirDAO.addGroupMember(allUsersGroup.id, defaultUser.id, samRequestContext)).thenReturn(IO(true))
+    when(dirDAO.removeGroupMember(allUsersGroup.id, defaultUser.id, samRequestContext)).thenReturn(IO(true))
+    when(dirDAO.isGroupMember(allUsersGroup.id, defaultUser.id, samRequestContext)).thenReturn(IO(true))
+    when(dirDAO.isEnabled(defaultUser.id, samRequestContext)).thenReturn(IO(true))
+    when(dirDAO.listUserDirectMemberships(defaultUser.id, samRequestContext)).thenReturn(IO(LazyList(allUsersGroup.id)))
+    when(dirDAO.setGoogleSubjectId(defaultUser.id, defaultUser.googleSubjectId.get, samRequestContext)).thenReturn(IO(()))
+    when(dirDAO.setUserAzureB2CId(defaultUser.id, defaultUser.azureB2CId.get, samRequestContext)).thenReturn(IO(()))
```
Comment (Ghost-in-a-Jar) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpec.scala:95]: maybe this block could be in a well-named fn just so you can understand what it is doing at a glance. I think that could be a followup pr though.

**Comment by gpolumbo-broad** [comment]
jenkins retest this please

**Pull request merged**